File: 05-BINARY-SEARCH/search-a-2d-matrix.py
from typing import *


# https://leetcode.com/problems/search-a-2d-matrix/
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        for row in matrix:
            if not (target >= row[0] or target <= row[-1]):
                continue

            low = 0
            high = len(row) - 1

            while low <= high:
                mid = (low + high) // 2

                if target > row[mid]:
                    low = mid + 1
                elif target < row[mid]:
                    high = mid - 1
                else:
                    return True

        return False

File: 05-BINARY-SEARCH/PHP/search-a-2d-matrix.php
<?php

# https://leetcode.com/problems/search-a-2d-matrix/
class Solution
{
    public function searchMatrix(array $matrix, int $target): bool
    {
        foreach ($matrix as $row) {
            if (!($target >= $row[0] || $target <= end($row))) {
                continue;
            }

            $low = 0;
            $high = count($row) - 1;

            while ($low <= $high) {
                $mid = floor(($low + $high) / 2);

                if ($target > $row[$mid]) {
                    $low = $mid + 1;
                } elseif ($target < $row[$mid]) {
                    $high = $mid - 1;
                } else {
                    return true;
                }
            }
        }

        return false;
    }
}